File: busy-beaver-scala/beaver.scala
import scala.collection.mutable.Buffer

abstract sealed class MachineMove(val value:Int)
case object Left extends MachineMove(-1)
case object Stay extends MachineMove(0)
case object Right extends MachineMove(1)

case class MachineTransition(write:Boolean, move:MachineMove, switch:Int)
case class MachineState(zero:MachineTransition, one:MachineTransition)
case class MachineDefinition(states:IndexedSeq[MachineState])

class Machine(val definition:MachineDefinition) {
    var state = 0
    var position = 0
    val tapeLeft = Buffer(false)
    val tapeRight = Buffer(false)

    override def toString():String = {
        s"state:$state position:$position tape:" +
        tapeLeft.reverseMap( if(_) '1' else '0' ).mkString +
        '|' + tapeRight.map( if(_) '1' else '0').mkString
    }

    def step {
        if(state == -1)
            return
        val (tape,index) = if(position>=0)
            (tapeRight,position)
        else
            (tapeLeft, -position-1)
        if(index >= tape.size) {
            tape += false
        }
        val read = tape(index)
        val s = definition.states(state)
        val transition = if(read) s.one else s.zero
        tape(index) = transition.write
        state = transition.switch
        position += transition.move.value
    }
}



object Beaver {
    def main(args: Array[String]) {
        val definition = MachineDefinition(Array(
            MachineState (
                MachineTransition (true,Right,1),
                MachineTransition (true,Left,2)),
            MachineState (
                MachineTransition (true,Right,2),
                MachineTransition (true,Right,1)),
            MachineState (
                MachineTransition (true,Right,3),
                MachineTransition (false,Left,4)),
            MachineState (
                MachineTransition (true,Left,0),
                MachineTransition (true,Left,3)),
            MachineState (
                MachineTransition (true,Right,-1),
                MachineTransition (false,Left,0))
        ))
        (0 until 100).foreach { index =>
            var i = 0
            val start = System.currentTimeMillis
            val machine = new Machine(definition)
            while(machine.state != -1) {
                machine.step
                i+=1
            }
            println(s"i:$i " + (System.currentTimeMillis - start));
        }
    }
}

File: busy-beaver-swift/beaver.swift
enum MachineMove:Int {
    case Left = -1, Stay = 0, Right = 1
}

struct MachineTransition {
    let write:Bool
    let move:MachineMove
    let switchTo:Int
}

struct MachineState {
    let zero:MachineTransition
    let one:MachineTransition
}

class MachineDefinition {
    let states:[MachineState]
    init(s:[MachineState]) {
        states = s
    }
}

private class Machine {
    let definition: MachineDefinition
    var state: Int = 0
    var position: Int = 0
    var tapeLeft:[Bool] = []
    var tapeRight:[Bool] = []

    init(def:MachineDefinition) {
        definition = def
    }

    func toString() -> String {
/*
        s"state:$state position:$position tape:" +
        tapeLeft.reverseMap( if(_) '1' else '0' ).mkString +
        '|' + tapeRight.map( if(_) '1' else '0').mkString
*/
        return "state:\(state) position:\(position) tape: \(tapeLeft) \(tapeRight)"
    }

    func step() {
        if(state == -1) {
            return
        }
        if(position >= 0) {
            if(position >= tapeRight.count) {
                tapeRight.append(false)
            }
            let read = tapeRight[position]
            let s = definition.states[state]
            let transition = read ? s.one : s.zero
            tapeRight[position] = transition.write
            state = transition.switchTo
            position = position + transition.move.rawValue
        } else {
            if(-position-1 >= tapeLeft.count) {
                tapeLeft.append(false)
            }
            let read = tapeLeft[-position-1]
            let s = definition.states[state]
            let transition = read ? s.one : s.zero
            tapeLeft[-position-1] = transition.write
            state = transition.switchTo
            position = position + transition.move.rawValue
        }
    }
}

let definition = MachineDefinition(s:[
    MachineState (
        zero:MachineTransition (write:true,move:MachineMove.Right,switchTo:1),
        one :MachineTransition (write:true,move:MachineMove.Left,switchTo:2)),
    MachineState (
        zero:MachineTransition (write:true,move:MachineMove.Right,switchTo:2),
        one :MachineTransition (write:true,move:MachineMove.Right,switchTo:1)),
    MachineState (
        zero:MachineTransition (write:true,move:MachineMove.Right,switchTo:3),
        one :MachineTransition (write:false,move:MachineMove.Left,switchTo:4)),
    MachineState (
        zero:MachineTransition (write:true,move:MachineMove.Left,switchTo:0),
        one :MachineTransition (write:true,move:MachineMove.Left,switchTo:3)),
    MachineState (
        zero:MachineTransition (write:true,move:MachineMove.Right,switchTo:-1),
        one :MachineTransition (write:false,move:MachineMove.Left,switchTo:0))
])
for index in 0...10 {
    var i = 0
    let machine = Machine(def:definition)
    while(machine.state != -1) {
        machine.step()
        i+=1
    }
}
